### PR TITLE
PHOENIX-6910 Scans created during query compilation and execution aga…

### PR DESCRIPTION
**Pull request by stoty**
…inst salted tables need to be more resilient

add test from PHOENIX-4096

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:1138]: here `currentGuidePostBytes` could be less than `endKey` right?

**Review by virajjasani — COMMENTED** [review]
Left minor comments, looks good overall

**Comment by virajjasani** [comment]
FYI @kadirozde @jpisaac @tkhurana you might be interested, this is resolving the scan boundary problem for salted tables after split/merge.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java:42]: shall we make this `SaltedTableSplitMergeBucketsIT` and add another test that does splitting of salted table region and then verify the result of SELECT queries against randomly picked key values from UPSERT queries?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java:100]: shall we pick random c1 values from a list of values instead of only with `20191211` and `20191217`?

we can create a list with multiple values and try using UPSERT in a loop against the list, that will reduce num of unique UPSERT statements we need to run in the test and also facilitate querying against any random c1 values from the given list.

Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java:100]: Random values in tests are fishy.
We could increase the number of rows, and check that we get back each of them.
It's a bit tricky, as the nature of hashing makes it hard to make sure that we cover all buckets.

Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java]: The now generates and checks 100 rows, that should be enough to hit all buckets (statistically).

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java]: Thanks, applied.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java]: Done

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java:47]: I have made the test more exhaustive, to make sure that we hit all buckets.
We could add a test for splits, but AFAIK there is no problem with splits.
This specific problem only manifests on merges.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java
Comment (stoty) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/salted/SaltedTableMergeBucketsIT.java]: Done.
Also added checks to make sure that we get the correct row.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java
Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:1138]: Thanks for pointing this out, @virajjasani .
This was actually a bug in my patch: 
Even if the new code split a scan because of crossing bucket boundaries, we still used the un-split region's key, which meant that we may have grouped different buckets together, which would have broken ordering.

I have removed the endKey parameter, and now use the new scan's end key.
I have also removed some logic that used the current region's start key as the end key, as it was unneccessary, and incorrect in the merged region case.

**Comment by stoty** [comment]
Hmm.
Even with my last change, @jpisaac 's test fail.
Looking into it.

**Comment by stoty** [comment]
I have added @jpisaac 's test, and fixed the error it deteceted.

**Comment by stoty** [comment]
The current version mostly works, but my refactoring has broken query statistics.
I hope to fix that later today.

**Comment by stoty** [comment]
Pushed fix for salted tables.
There is still a local index regression I am chasing.

**Comment by stoty** [comment]
All relevant test pass now.
I think that the FailoverPhoenixConnectionIT is unrelated.
Can you take a look now @virajjasani @jpisaac @kadirozde @tkhurana ?

**Comment by stoty** [comment]
Hmm.
I think my changes are confusing the start and end of region/bucket flags.
Gonna review and push a fix.
If you haven't started reviewing yet, hold off for a bit.

**Comment by stoty** [comment]
Now it should be really ready.
The stats=enabled case wasn't tested, and was broken.

**Comment by stoty** [comment]
Can you review this now that hopefully both cases are working @jpisaac , @virajjasani  ?

**Comment by virajjasani** [comment]
somehow jenkins didn't trigger new build since May 25, some issues with build maybe.

just triggered a new one [here](https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1607/)

**Comment by stoty** [comment]
That CI run has hung.
I'm starting another.
(We have a number of unresolved flakey tests related to the HA feature)

**Comment by virajjasani** [comment]
oops, it seems some tests are taking longer and hence the build is timing out. 
@stoty do you have any vm or docker setup where you think you could run mvn verify and keep it running for a bit longer? because in this case, we might be able to get better insights from local runs i think.

**Comment by stoty** [comment]
If a test run doesn't finish in 3-4 hours then it never will.
I've had trouble running tests on either too fast or too slow machines.
Starting a local test now.

**Comment by stoty** [comment]
I have attached local-verify-run.gz to the ticket.
It only fails on the known flakey permission ITs.
(It seems that we have tests that tend to fail on slow machines, and also ones that tend to fail on fast ones)

**Comment by virajjasani** [comment]
Finally build runs were successful, let me take a look tomorrow.

At high level, IIUC, ParallelScansCollector is going to keep adding scan objects and all places where we use parallel scans, we are going to use this new class. New version of intersectScan is going to include scan boundaries from salt buckets. Is this correct?

**Comment by stoty** [comment]
Yes, that's correct.
ParallelScansCollector just simplifies the messy logic where we had to split the lists at the next entry, and maintain the state in object variables.
I have also consolidated some of the very dispersed logic into fewer places.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:158]: which query will be impacted due to the override?

**Review by virajjasani — COMMENTED** [review]
looks good overall, left minor comments

**Comment by virajjasani** [comment]
surprising that end2end tests were not run https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1607/28/testReport/

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java
Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:158]: None.
If this were fixed, we could take use the salting info from the scanRange object.
Now, we have to go back to the PTable to get it, and the salting info in ScanRange is not used.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/ParallelScansCollector.java
Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/ParallelScansCollector.java:34]: Will fix in a moment

**Comment by stoty** [comment]
I'm pretty sure that the test were run. 
We're just aging them out really quick, because the Jenkins workers have very little disk space.
The final changes will trigger a new CI run.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java
Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:158]: Currently none.
We just need to re-read some salting info from the table, instead of taking the cached value from

**Comment by stoty** [comment]
I can see that there are also quite a few warnings.
I will address (some of) those based on the new CI run.

**Comment by stoty** [comment]
I took a look at the warning, they are mostly for existing issues.
I do not plan to fix those, as that would only add more noise to the patch.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:158]: got it, sounds good

**Review by virajjasani — APPROVED** [review]
+1

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java
Comment (richardantal) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:42]: Can you remove this import please or list everything needed under java.util?

Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:42]: You are right, but this has already been committed.
Can you open a follow-up ticket ?

**Pull request closed without merging**
